Building virtio example with zig

erichchan999 opened the pull request:
Waiting on this PR #74 to be merged first.

Provides zig build for virtio example. Changes the top-level build.zig in libvmm to have sddf build.zig as a dependency.

Ivan-Velickovic committed:
manual: improvements, additions, fixes

* Various clarifications and fixes
* Adding a new section on pass-through devices
* Making the styling of the manual nicer

Ivan-Velickovic committed:
examples/zig: fixing and cleaning up build script

Ivan-Velickovic committed:
examples/zig: fixes and cleanup in VMM to get Linux to boot

For some reason building in ReleaseFast mode, `-Doptimize=ReleaseFast`
successfully boots Linux and lets us do serial input, but not debug
mode.

More investigation required.

Ivan-Velickovic committed:
examples/zig: add more detail to README

Ivan-Velickovic committed:
examples/zig: remove debug/commented code

Ivan-Velickovic committed:
examples/zig: more details in README for running the example

Ivan-Velickovic committed:
examples/zig: more cleanup in the VMM code

Ivan-Velickovic committed:
examples/zig: cleaning up build.zig

Ivan-Velickovic committed:
examples/zig: adding Zig example to CI

Ivan-Velickovic committed:
libvmm: add more error-checking to guest images setup

Ivan-Velickovic committed:
libvmm: make a small note about Linux images setup API

Ivan-Velickovic committed:
libvmm: missing include in DTB header

`stdbool.h` is technically missing, it's already included by users
of `dtb.h` which is why we have not seen a compiler error for it yet.
But, better to fix this now than later.

Ivan-Velickovic committed:
examples/zig: do not strip symbol table from ELF

Ivan-Velickovic committed:
Add support for Avnet MaaXBoard

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Get GICv3 driver into a compiling state

Ivan-Velickovic committed:
README: mention what VMM stands for

Ivan-Velickovic committed:
README: mention roadmap

Ivan-Velickovic committed:
lib: log error when vIRQ cannot be set to pending

These logs should be errors rather than just normal logs as it is
a case that should not happen.

Ivan-Velickovic committed:
examples/virtio: init

This is an example showing off the virtIO support of libvmm.
It currently only shows off the virtIO console device, with more
devices such as block and graphics to be added later.

Ivan-Velickovic committed:
lib: add implementation of virtIO MMIO

Co-authored-by: Jingyao Zhou <jingyao.zhou@unsw.edu.au>

Ivan-Velickovic committed:
lib: add implementation of virtIO console device

Ivan-Velickovic committed:
examples/zig: update for latest Zig

0.12.0-dev.1533+b2ed2c4d4 was used.

Ivan-Velickovic committed:
examples/zig: turn off UBSAN

It's unclear whether we actually do have UB somewhere in the C code. For
now we turn off UBSAN in order to have the example work in debug mode.

Ivan-Velickovic committed:
examples/simple: update Zig version used

Ivan-Velickovic committed:
examples/simple: install DTB into build directory

The build system was confusing and I couldn't get this working and gave
up. Now with some new build system documentation I could figure it out.

Ivan-Velickovic committed:
examples/zig: move images/ out of src/ and minor clean up

Having 'src/images/' was done before-hand since I wasn't familiar with
how to include files from 'images' in the VMM if they were outside the
'src' directory.

Ivan-Velickovic committed:
examples/simple: minor clean up in build.zig

Ivan-Velickovic committed:
examples/zig: add more comments in VMM

Ivan-Velickovic committed:
examples/zig: remove unused struct

Ivan-Velickovic committed:
examples/zig: fix debug printing

Ivan-Velickovic committed:
Add LICENSE

Ivan-Velickovic committed:
examples/virtio: update sDDF module to new branch

Ivan-Velickovic committed:
Add manual dependencies to Nix shell

Ivan-Velickovic committed:
examples/zig: updating libmicrokit.zig

0.12.0-dev.1664+8ca4a5240 caused some breaking changes and so we
have to re-create libmicrokit.zig

Ivan-Velickovic committed:
examples/zig: add documentation for creating libmicrokit.zig

Mainly for my own sake since I am very forgetful these days.

Ivan-Velickovic committed:
examples/zig: minor styling and fixes in README

Ivan-Velickovic committed:
manual: add dependencies to README

Ivan-Velickovic committed:
manual: remove unused image

Ivan-Velickovic committed:
lib: add missing white-space in error messages

Ivan-Velickovic committed:
lib: fix error printing

alexandermbrown committed:
virtio-snd: add virtIO device and UIO driver

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

Ivan-Velickovic committed:
examples/virtio: update sDDF module used

The `lionsos` branch no longer exists.

Ivan-Velickovic committed:
Update verison of Microkit SDK to download in README

Ivan-Velickovic committed:
Change README instructions for acquiring SDK

The artefacts produced from the GitHub Actions expire after
a certain amount of time and require you to be logged in. This
is easier for the meantime while we still need a dev version of
Microkit.

Ivan-Velickovic committed:
Adjusting wording in various READMEs and comments

Ivan-Velickovic committed:
examples/simple: don't strip symbol table for Zig build

Ivan-Velickovic committed:
lib: remove TODO

There is no initrd magic to check since the initrd could be a variety
of file formats.

Ivan-Velickovic committed:
Fix README instructions

Ivan-Velickovic committed:
Fix wording when referring to libvmm

Ivan-Velickovic committed:
More README cleanup

Ivan-Velickovic committed:
Update package used for TexLive in Nix shell

I believe the Nix package manager recently renamed the full
TexLive package.

Ivan-Velickovic committed:
README typo fix

Ivan-Velickovic committed:
examples/zig: README fixes

Ivan-Velickovic committed:
Fixes for latest master Zig version

There are some minor breaking API changes to the Zig build system. It
is good to get ahead of these so we don't have to deal with all of them
at the next release of Zig.

Hopefully soon we can pin the examples that use Zig to an official
release rather than the master version.

Ivan-Velickovic committed:
ci: minor cleanup

Ivan-Velickovic committed:
examples/simple: do not hard-code board flag in build.zig

Ivan-Velickovic committed:
examples/simple: disable UBSAN when building with Zig

Zig enables UBSAN on C code by default and it seems to be creating
weird behaviour. If UB is detected at run-time, the program should
panic by executing a `brk` instruction in this case (ARM64 and no OS).

It seems that the Microkit event loop is actually doing different
things depending on whether or not UBSAN is on. Specifically, the
`seL4_NBSendRecv` system call is being made when UBSAN is on, even
though it should never be called in this example code (and is not
called when we turn UBSAN off).

Not sure what's going on, it's still entirely possible that we have
UB in our code. More investigation required.

Ivan-Velickovic committed:
Update LLVM tools version in Nix shell

alexandermbrown committed:
Fix virtIO MMIO QueueNotify field

Ivan-Velickovic committed:
examples/virtio: use colour in console output

Ivan-Velickovic committed:
ci: fix upload of manual PDF

Ivan-Velickovic committed:
examples/virtio: fix for latest sDDF

Some minor breaking changes with includes and defines

Ivan-Velickovic committed:
ci: update deprecated GitHub action versions

Ivan-Velickovic committed:
Update Microkit SDK version in README

Ivan-Velickovic committed:
Minor docs fixes

Ivan-Velickovic committed:
ci: temporarily disable testing on macOS via Nix

For some reason the Rust example does not work on macOS using
a Nix environment. Not sure why as it works for everything other
environment we test. Will investigate later.

The logs are:
     Updating git repository `https://github.com/Ivan-Velickovic/rust-seL4`
    Updating crates.io index
warning: spurious network error (3 tries remaining): [35] SSL connect error (OpenSSL/3.0.12: error:16000069:STORE routines::unregistered scheme)
warning: spurious network error (2 tries remaining): [35] SSL connect error (OpenSSL/3.0.12: error:16000069:STORE routines::unregistered scheme)
warning: spurious network error (1 tries remaining): [35] SSL connect error (OpenSSL/3.0.12: error:16000069:STORE routines::unregistered scheme)
error: failed to get `cfg-if` as a dependency of package `sel4-microkit v0.1.0 (https://github.com/Ivan-Velickovic/rust-seL4?branch=dev#e4bde410)`
make: Leaving directory '/Users/runner/work/libvmm/libvmm/examples/rust'
    ... which satisfies git dependency `sel4-microkit` (locked to 0.1.0) of package `vmm v0.1.0 (/Users/runner/work/libvmm/libvmm/examples/rust)`

Caused by:
  failed to query replaced source registry `crates-io`

Ivan-Velickovic committed:
tools: add dtscat script

Ivan-Velickovic committed:
Fix typo in comment

Ivan-Velickovic committed:
Do not write out all of seL4_UserContext when starting guest

Ivan-Velickovic committed:
Minor style fixes/removing commented code

Ivan-Velickovic committed:
Minor cleanup in fault decoding

Ivan-Velickovic committed:
Actually turn asserts off in debug build

Can't believe this has been like this for this long.

Ivan-Velickovic committed:
Fix incorrect assumption in GIC driver

Before handling an IRQ, we should first be checking that it has
been registered with the GIC by the VMM.

Ivan-Velickovic committed:
Add API for registering passthrough IRQs

This API is mainly for convenience, it does not do anything that
people could not already do but makes the implementations of VMMs
simpler.

See the comments in the header file for details.

Co-authored-by: Jingyao Zhou <jingyao.zhou@unsw.edu.au>

Ivan-Velickovic committed:
Restructuring the Zig build system

This gives libvmm a top-level build.zig which means that projects that
depend on libvmm can simply add libvmm as a dependency in their
build.zig.zon file and do not have to figure out how to build libvmm.a.

This same approach is used by some of the examples.

Ivan-Velickovic committed:
examples/rust: update rust-seL4 version

Necessary since there was a change in the Microkit SDK that
affects the Rust stuff. We're using a fairly old version of
rust-seL4 hence why we need to update.

Toolchain needs to be updated as well due to use of unstable
features.

wom-bat committed:
Fix quoting

Use "$@" not "$*" so that spaces in arguments are preserved.
Also add the #! tag at the start.

Signed-off-by: Peter Chubb <Peter.Chubb@unsw.edu.au>

wom-bat committed:
Add back extra blank line

As requested.

Signed-off-by: Peter Chubb <Peter.Chubb@unsw.edu.au>

alexandermbrown committed:
manual: add DMA passthrough section (#39)

wom-bat committed:
Serial queue (#51)

* Update mmio.h to match serial queue renaming

sharedringbuffer.h has been renamed queue.h

Signed-off-by: Peter Chubb <Peter.Chubb@unsw.edu.au>

* Update for sDDF serial queue renaming

Everywhere rename ring->queue, used->active

Signed-off-by: Peter Chubb <peter.chubb@unsw.edu.au>

* Updated submodule sDDF to head, to get queue renaming changes

Signed-off-by: Peter Chubb <peter.chubb@unsw.edu.au>

* Make the Makefile use the sDDF submodule

Signed-off-by: Peter Chubb <Peter.Chubb@unsw.edu.au>

* More virtio example fixes

Rename MUX->Virtualiser, ring->queue
Fix Makefile to track dependencies properly,
Fix printf.c name clash (the uart driver wants the sddf version, vmm.o
wants the libvmm version

Signed-off-by: Peter Chubb <Peter.Chubb@unsw.edu.au>

---------

Signed-off-by: Peter Chubb <Peter.Chubb@unsw.edu.au>
Signed-off-by: Peter Chubb <peter.chubb@unsw.edu.au>

wom-bat committed:
Add Makefile snippet for libvmm.a (#46)

* Add Makefile snippet for libvmm.a

To use this, in your project Makefile add:

LionsOS := <path-to-top-of-LionsOS-tree>
vmm: libvmm.a
include ${LionsOS}/vmm.mk

Signed-off-by: Peter Chubb <peter.chubb@unsw.edu.au>

* Don't require knowledge of LionsOS

Instead allow the SDDF variable to be set by an including Makefile, or
(_if_ vmm is being build as a LionsOS component) infer the value of
SDDF

Signed-off-by: Peter Chubb <Peter.Chubb@unsw.edu.au>

* Fix typo

Include auto-generated dependencies properly.

Signed-off-by: Peter Chubb <Peter.Chubb@unsw.edu.au>

* Add clean and clobber targets

Signed-off-by: Peter Chubb <Peter.Chubb@unsw.edu.au>

* Get rid of LionsOS mentions

Always require SDDF variable to be set before including this snippet.

Signed-off-by: Peter Chubb <Peter.Chubb@unsw.edu.au>

---------

Signed-off-by: Peter Chubb <peter.chubb@unsw.edu.au>
Signed-off-by: Peter Chubb <Peter.Chubb@unsw.edu.au>

Ivan-Velickovic committed:
Update to latest Zig version

Breaking change to Zig standard library. Once Zig 0.12.0
will release we will make libvmm use that, but for now
we stick to master.

wom-bat committed:
Add documention for the libvmm interfaces (#50)

* Add documention for the libvmm interfaces

Signed-off-by: Peter Chubb <Peter.Chubb@unsw.edu.au>

* Move interface description into docs directory

Signed-off-by: Peter Chubb <peter.chubb@unsw.edu.au>

* Review changes

Signed-off-by: Peter Chubb <Peter.Chubb@unsw.edu.au>

* Minor fix

Signed-off-by: Peter Chubb <Peter.Chubb@unsw.edu.au>

---------

Signed-off-by: Peter Chubb <Peter.Chubb@unsw.edu.au>
Signed-off-by: Peter Chubb <peter.chubb@unsw.edu.au>

wom-bat committed:
Make libvmm depend on CFLAGS

When the microkit BOARD or CONFIG change, we want to force rebuilding
of the library, to pick up changes in the header files.

Signed-off-by: Peter Chubb <Peter.Chubb@unsw.edu.au>

alexandermbrown committed:
virtio-snd: add virtIO support level to docs

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: virtIO refactor

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

erichchan999 committed:
Format to CI style

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

erichchan999 committed:
Fixing headers related to virtio

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

erichchan999 committed:
Update virtio console comments to sddf rename changes

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

erichchan999 committed:
virtio mmio refactor

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

erichchan999 committed:
Make memcpy and memset externally linked

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

erichchan999 committed:
Add printing in hex convenience function

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

erichchan999 committed:
add virtio block implementation

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

erichchan999 committed:
add libuio and uio blk driver implementation

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

erichchan999 committed:
Add script to generate virtual disk with mbr/msdos partition table

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

erichchan999 committed:
Add virtio block system to virtio example

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

alexandermbrown committed:
virtio-snd: add minor sDDF sound improvements

https://github.com/au-ts/sddf/pull/83

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

Ivan-Velickovic committed:
ci: test to see if installing gcc-aarch64-linux-gnu works

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

alexandermbrown committed:
virtio-snd: use busy wait on init

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: remove q size from shared memory

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: switch back to freelist

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

abrandnewusername committed:
Implement virtIO net device

Co-authored-by: Alex Brown <alexandermichaelbrown@outlook.com>
Signed-off-by: Jingyao Zhou <jingyao.zhou@unsw.edu.au>

Ivan-Velickovic committed:
ci: build for Odroid-C4 in virtIO example

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
ci: regress virtIO example on macOS temporarily

In order to merge the virtIO block PR, we will sort out
getting it working on macOS later.

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
nix: add aarch64-unknown-linux-gcc

Needed for virtIO example since it cross-compiles for Linux
user-space.

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

erichchan999 committed:
Add handling for block size

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

erichchan999 committed:
Check return for fsync

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

erichchan999 committed:
Check UIO IRQ injection

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

erichchan999 committed:
Add comment explaining why we need sddf_max_data_buffers

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

erichchan999 committed:
Clean up ericc commennts

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

erichchan999 committed:
Update README to have max 100 character lines

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

erichchan999 committed:
Update README

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

erichchan999 committed:
Better commenting on libuio driver functions

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

erichchan999 committed:
Various virtIO example/block changes and fixes

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

alexandermbrown committed:
virtio-snd: improve buffer addressing

- Make client buffer addresses relative to base
- Make driver buffer addresses host-physical
- Driver VM now converts host-physical to guest-virtual by passing
  host-physical address of data region to guest through shared memory.

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: rename data_*addr to data_region_*addr

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: remove .clang-format

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: label passthrough IRQs

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: remove unused min function

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: fix Makefile after rebase

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
Update virtIO block library

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
Add 3rd serial client to examples/virtio

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
Concatenate rootfs instead of unpacking

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: move uio_sound.h to tools

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: remove cantina & crav_rave

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: add comment to alsa config

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: fix switch statement indentation

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: cleanup shell.nix

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
Make ialloc 32-bit and use in virtIO sound

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

Ivan-Velickovic committed:
sound: minor style changes

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

alexandermbrown committed:
virtio-snd: remove unused buffer.h

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
Add back virtIO 1.2 comment

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: add QUEUE_BYTES comment

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

Ivan-Velickovic committed:
Introduce 'dep' dir and put sDDF in there

Better than having a submodule in the example
directory.

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

alexandermbrown committed:
virtio-snd: remove unused queue_name

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: rename SERIAL_MUX_*_CH

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: change sddf directory

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: remove vgic include

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: remove poll.c

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: move SOUND.md to virtio-snd

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

Ivan-Velickovic committed:
virtio: add some comments to MMIO functions

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
sound: fixup README for virtIO sound example

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Minor README consistency

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

alexandermbrown committed:
virtio-snd: use packrootfs

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: remove incorrect license

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: comment out DEBUG_SOUND

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: remove rootfs_overlayed from gitignore

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: docs: use svg instead of png

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: remove @alexbr comments

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

Ivan-Velickovic committed:
virtio-snd: restructure docs a bit

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
examples/virtio: remove shell.nix

Not necessary for this example

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

alexandermbrown committed:
virtio-snd: coalesce userlevel programs

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: remove pack.sh

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
Remove old sddf_handlers from mmio.h

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

Ivan-Velickovic committed:
examples/virtio-snd: remove whitespace

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

alexandermbrown committed:
virtio-snd: docs: remove cantina.elf

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
virtio-snd: docs: update startup to busy wait

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

Ivan-Velickovic committed:
virtio: remove more whitespace

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
virtio-snd: remove unecessary include

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Remove unecessary headers

We don't access the vGIC directly, only the vIRQ
API.

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

erichchan999 committed:
mkvirtdisk refactor

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

wom-bat committed:
virtio.c is gone: update vmm.mk to match

Signed-off-by: Peter Chubb <peter.chubb@unsw.edu.au>

Ivan-Velickovic committed:
Update to latest Zig version

Again there's some breaking changes that affect us.

We should be able to pin to 0.13.0 of Zig which will be the next
release but at the moment there's a couple patches we want that
aren't in the current latest release so we have to stick to master.

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

wom-bat committed:
Add block.c and sound.c to the virtio drivers to add to the VMM

Signed-off-by: Peter Chubb <peter.chubb@unsw.edu.au>

Ivan-Velickovic committed:
Add licenses to comply with REUSE specification

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Use normal nixpkgs for virtio-snd example

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Fix for latest Zig

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Courtney3141 committed:
Update blk driver to use atomic store

Signed-off-by: Courtney Darville <courtneydarville94@outlook.com>

Courtney3141 committed:
Update libvmm to use latest sDDF util and serial system

Signed-off-by: Courtney Darville <courtneydarville94@outlook.com>

Courtney3141 committed:
Fix typos

Signed-off-by: Courtney Darville <courtneydarville94@outlook.com>

Courtney3141 committed:
Update to latest sDDF

Signed-off-by: Courtney Darville <courtneydarville94@outlook.com>

Ivan-Velickovic committed:
Pin to version 0.13.0 of Zig

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

erichchan999 committed:
add set -e to mkvirtdisk

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

erichchan999 committed:
Change ar and ranlib to use llvm's

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

erichchan999 committed:
Typo

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

erichchan999 committed:
Move SERIAL_NUM_CLIENTS to serial_config.h

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>

Ivan-Velickovic committed:
Fix typos in build.zig files

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Move to Microkit 1.4.0

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

erichchan999 committed:
Upstream sddf blk to use offsets, and initialise blk system with blk_config.h

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>
Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Update sDDF submodule

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Small style fixes

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
examples/virtio: remove whitespace

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Cleanup to virtIO console device

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
examples/virtio: README cleanup

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Update sDDF submodule

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
examples/virtio-snd: do not depend on GCC AR/ranlib

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
examples/virtio-snd: add extra nix dep

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Minor makefile cleanup

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
build.zig: improvements

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic commented on build.zig:
```diff
@@ -25,17 +25,36 @@ const src_aarch64 = [_][]const u8{
     "src/arch/aarch64/vcpu.c",
 };
 
+const src_virtio = [_][]const u8{
+    "src/virtio/block.c",
+    "src/virtio/console.c",
+    "src/virtio/mmio.c",
+    "src/virtio/sound.c",
+};
```
Put these in `src` since they'll always be compiled.

erichchan999 committed:
Zig build for virtio example

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>
Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

erichchan999 committed:
Rename uio_blk_driver -> uio_driver_blk

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>
Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

erichchan999 committed:
Add set -e to scripts

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>
Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

erichchan999 committed:
Pipe tools stderr to /dev/null to silence zig build errors

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>
Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

erichchan999 committed:
Remove problematic stdout -l print in fdisk

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>
Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Sort out our includes

This commit puts all includes in their own directory.

This is done to simplify building and using libvmm such that
there is only one include directory to install/add to path.

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

alexandermbrown committed:
vmm.mk: remove BOARD prefix, add maaxboard as V3 board

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
examples/simple: refactor Makefile to use snippets

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>
Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

alexandermbrown committed:
vmm.mk: use shasum instead of md5

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

Ivan-Velickovic committed:
ci: remove --pure from Nix invocation

Not needed since we don't set any envars. The motiviation is that
we need shasum which is generally available on major Linux distros
pre-installed but not on Nix.

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic commented on examples/virtio/build.zig.zon:
```diff
@@ -0,0 +1,17 @@
+            .path = "../../dep/sddf"
+        }
+    },
+    .paths = .{
```
The example isn't supposed to be a package that other people use so we should just leave paths empty `.paths = {}`.

erichchan999 replied:
an empty string does the some thing i believe, zig is not happy when you leave the dictionary empty

Ivan-Velickovic replied:
Nevermind then, whoops.

Ivan-Velickovic commented on examples/virtio/build.zig.zon:
```diff
@@ -0,0 +1,17 @@
+.{
+    .name = "virtio",
```
```suggestion
    .name = "virtio_example",
```

Ivan-Velickovic commented on build.zig:
```diff
@@ -72,10 +88,49 @@ pub fn build(b: *std.Build) void {
 
     // @ivanv: fix all of our libvmm includes! This is a mess!
     libvmm.addIncludePath(b.path("src"));
-    libvmm.addIncludePath(b.path("src/util/"));
+    libvmm.addIncludePath(b.path("src/util"));
+    libvmm.addIncludePath(b.path("src/virtio"));
     libvmm.addIncludePath(b.path("src/arch/aarch64"));
-    libvmm.addIncludePath(b.path("src/arch/aarch64/vgic/"));
+    libvmm.addIncludePath(b.path("src/arch/aarch64/vgic"));
```
Once you rebase on main we can remove all these lines, finally.

Ivan-Velickovic commented on build.zig:
```diff
@@ -72,10 +88,49 @@ pub fn build(b: *std.Build) void {
+    libvmm.linkLibrary(sddf_dep.artifact("util"));
+    libvmm.linkLibrary(sddf_dep.artifact("util_putchar_debug"));
+
+    libvmm.installHeadersDirectory(b.path("src"), "", .{});
+    libvmm.installHeadersDirectory(b.path("src/util"), "", .{});
+    libvmm.installHeadersDirectory(b.path("src/virtio"), "", .{});
+    libvmm.installHeadersDirectory(b.path("src/arch/aarch64"), "", .{});
+    libvmm.installHeadersDirectory(b.path("src/arch/aarch64/vgic"), "vgic", .{});
```
Same thing here.

Ivan-Velickovic commented on build.zig:
```diff
@@ -72,10 +88,49 @@ pub fn build(b: *std.Build) void {
 
     b.installArtifact(libvmm);
+
+    var target_userlevel_query = target.query;
+    target_userlevel_query.os_tag = .linux;
+    target_userlevel_query.abi = .musl;
+    const target_userlevel = b.resolveTargetQuery(target_userlevel_query);
```
This is a bit weird, we should just construct a new target query instead of copying and modifying the current one.

erichchan999 replied:
i had in mind that we don't want the outerlevel zig build to know anything about platform specific stuff. otherwise we'll need to pass in the board we're using, or the specific architecture we're working with to construct that target. in the end we only care about changing the os_tag and abi anyway and not the architecture. happy to do either

erichchan999 committed:
Refactor disables from the base linux.dts into disable.dts in qemu, as is done for odroidc4 example

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>
Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

erichchan999 committed:
Update virtio example disable.dts changes to zig build

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>
Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic commented on examples/virtio/build.zig:
```diff
@@ -0,0 +1,498 @@
+    // Microkit tool command
+    const system_description_path = LazyPath{ .cwd_relative = b.fmt("board/{s}/virtio.system", .{ @tagName(microkit_board) }) };
+    const final_image_dest = b.getInstallPath(.bin, "./loader.img");
+    const microkit_tool_cmd = b.addSystemCommand(&[_][]const u8{
+        microkit_tool.getPath(b),
+    });
+    microkit_tool_cmd.addFileArg(system_description_path);
+    microkit_tool_cmd.addArg("--search-path");
+    microkit_tool_cmd.addArg(b.getInstallPath(.bin, ""));
+    microkit_tool_cmd.addArg("--board");
+    microkit_tool_cmd.addArg(@tagName(microkit_board));
+    microkit_tool_cmd.addArg("--config");
+    microkit_tool_cmd.addArg(@tagName(microkit_config));
+    microkit_tool_cmd.addArg("-o");
+    microkit_tool_cmd.addArg(final_image_dest);
+    microkit_tool_cmd.addArg("-r");
+    microkit_tool_cmd.addArg(b.getInstallPath(.prefix, "./report.txt"));
+    microkit_tool_cmd.step.dependOn(b.getInstallStep());
```
Can you follow what `examples/simple/build.zig` does for this?

Ivan-Velickovic commented on examples/virtio/build.zig:
```diff
@@ -0,0 +1,498 @@
+        mkvirtdisk_option = b.option(bool, "mkvirtdisk", "Create a new virtual disk image") orelse false;
+    }
+
+    // Libmicrokit
```
```suggestion
    // libmicrokit
```

Ivan-Velickovic commented on examples/virtio/build.zig:
```diff
@@ -0,0 +1,498 @@
+    // Create virtual disk if it doesn't exist or if the user wants to create a new one
+    if (mkvirtdisk_option or !fileExists(b.getInstallPath(.prefix, virtdisk_name))) {
+        // TODO: Make these configurable options
+        const num_part = "2";
+        const logical_size = "512";
+        const blk_mem = "2101248";
```
These should all be integers and converted to strings in `mkvirtdistCmd`.

Ivan-Velickovic committed:
ci: add Zig build for virtio example

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
examples/virtio: rename in build.zig.zon

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Remove trailing whitespace

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
ci: remove redundant todo

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
ci: add style and trailing whitespace check

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

erichchan999 committed:
Various fixes

Signed-off-by: Eric Chan <eric.chan2@unsw.edu.au>
Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
test commit to see if CI passes now

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
ci: fix, oopsies

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
build.zig: cleanup

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
More cleanup

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Cleanup

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

alexandermbrown committed:
examples/virtio: use Makefile snippets

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

Ivan-Velickovic committed:
Remove 'framebuffer' and 'ethernet' examples

These examples are unmaintained and not tested in CI. What the
examples show is worthwhile however and is planend to put in a
'passthrough' example.

I don't know when we will get to creating that passthrough example,
so in the interest of not having bitrotted examples, we remove them.

They are of course in the commit history if we ever want to revive
them.

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Implement virtio console reset

A trivial patch and gets rid of the
stupid 'virtio_console_reset is not implemented' message that has
been around for a long time whenever we boot the virtio example.

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Use false instead of 0 in virtIO MMIO reset

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

alexandermbrown committed:
examples/virtio-snd: use Makefile snippets

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

Ivan-Velickovic committed:
Remove outdated comment in vmm.mk

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

alexandermbrown committed:
examples/simple: remove bashism

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
examples/virtio: remove client 2 files

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
examples/virtio-snd: fix odroid serial mappings

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

Ivan-Velickovic committed:
examples/virtio: update README

Commit 5628ccdc49368a40c92dcc8c8c7d92ad11d3c557 changed the images
to be the same so this note needs to be updated.

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
manual: add docs for virtIO

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

alexandermbrown committed:
Update serial config to new syntax

The latest version of sDDF changes how `serial_virt_queue_init_sys`
is defined in `serial_config.h`. This change allows the latest version
of sDDF serial to be used.

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
blk: rename size to capacity

This is required for libvmm to compile with latest version of sDDF

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

alexandermbrown committed:
Update sDDF commit

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

Ivan-Velickovic committed:
Fix style for virtio/block.c and uio_tools/blk.c

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Remove outdated comments

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Use bool for callbacks in virtIO devices

We already actually cast the return values to 'bool' in `src/mmio.c`
anyways, this just makes things a bit more consistent.

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Move macro definition util.h

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
style: virtIO console.c and console.h

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
style: virtIO sound.c and sound.h

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

wom-bat committed:
Make -I from SDDF and LIBVMM conditional

Only add them if they're not already there, to
allow higher level makefiles to adjust their order if necessary.

Plus move objects into a subdir called 'libvmm' instead of 'src' ---
the latter is too general.

Signed-off-by: Peter Chubb <Peter.Chubb@unsw.edu.au>

wom-bat committed:
Add missing space

Signed-off-by: Peter Chubb <Peter.Chubb@unsw.edu.au>

Ivan-Velickovic committed:
Move manual dependencies into separate Nix shell

They are large and not necessary for most people who are just
building the library/examples.

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
README: fix typo

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
README: don't enforce --pure for Nix

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Various README fixes

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

alexandermbrown committed:
examples/virtio-snd: fix VCPU for Microkit 1.4.0

Signed-off-by: Alex Brown <alexandermichaelbrown@outlook.com>

Ivan-Velickovic committed:
ci: add license check

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Update to sDDF release 0.5.0

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
shell.nix: fixes/additions

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Delete file from rebase

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
build.zig: fix for rebase

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
ci: build virtIO example for QEMU

There's an issue with Microkit 1.4.0 that affects the Odroid-C4.

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Remove file after rebase

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
License fixes

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Revert rename of UIO block driver

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
Remove TODO

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>

Ivan-Velickovic committed:
cleanup

Signed-off-by: Ivan Velickovic <i.velickovic@unsw.edu.au>
